### PR TITLE
Display invalid filenames as warning

### PR DESCRIPTION
**Pull request by FlexW**
This pr displays a warning in the activity log if a file could not be created because of illegal characters. It looks like this:

warnings

Additionally it shows an operating system notification.

The pr is quick and dirty at the moment. Few things to consider:

- Do we want to group the warnings? Like for ignored files?
- Should an operating system notification only displayed once when the issue first gets detected? 
  (At the moment with each sync run the notification gets displayed again)

Fixes: #3013, #3018, #464

### CODE REVIEW

File: src/gui/tray/UserModel.cpp [hunk 1]
```diff
@@ -486,6 +487,9 @@ void User::slotItemCompleted(const QString &folder, const SyncFileItemPtr &item)
             _activityModel->addIgnoredFileToList(activity);
         } else {
             // add 'protocol error' to activity list
+            if (item->_status == SyncFileItem::Status::FileNameInvalid) {
+                emit guiLog(item->_file, activity._subject);
```
Comment (allexzander) [inline src/gui/tray/UserModel.cpp:491]: @FlexW Not sure, but, I think this should be moved into `void User::slotBuildNotificationDisplay(const ActivityList &list)` as the other `emit guiLog()` is already there. Just to keep this part clean of redundant code :)

**Comment by allexzander** [comment]
@FlexW Looks good overall. Just 1 comment :)

**Comment by FlexW** [comment]
@allexzander Thanks. What do you think about the two questions that I asked?:)

**Comment by allexzander** [comment]
> @allexzander Thanks. What do you think about the two questions that I asked?:)

@FlexW Good idea. 
- Having different icons for different types of errors would definitely improve the UX, but, probably needs to be discussed with the design team.
- Yes, this is also a good idea. Having a caching mechanism in runtime would help to reduce spam. Also, should be easy to implement. Then, when the app is exited and started again - the notification will get displayed again. I think this is fine and not too spammy.

File: src/gui/tray/UserModel.cpp (hunk 1 above)
Comment (er-vin) [reply src/gui/tray/UserModel.cpp:491]: Yep, agreed with this. If that's doable it's better.

**Comment by FlexW** [comment]
@allexzander I cached the notifications so that they will only show up once.

File: src/gui/tray/UserModel.h [hunk 2]
```diff
@@ -91,11 +92,16 @@ public slots:
...
+    int notificationId(const QString &subject, const QString &message);
+    void displayNotification(const QString &title, const QString &message);
+    void displayNotification(int id, const QString &title, const QString &message);
+
 private:
     AccountStatePtr _account;
     bool _isCurrentUser;
     ActivityListModel *_activityModel;
     ActivityList _blacklistedNotifications;
+    QHash<QString, int> notificationIds;
```
Comment (er-vin) [inline src/gui/tray/UserModel.h:104]: Should start with _

**Comment by er-vin** [comment]
Also: is it me or that caching mechanism to avoid spamming notifications is a bit indiscriminate? I mean it doesn't take the type of the notification into account and I suspect there are some that you want to report at each sync because they are important enough for that.

**Comment by FlexW** [comment]
@er-vin I think this function was so far just used for activities and they were always be cashed. So I think it's fine. In the future, we can extend that if we need it.

File: src/gui/tray/UserModel.cpp [hunk 3]
```diff
@@ -496,11 +499,26 @@ void User::slotItemCompleted(const QString &folder, const SyncFileItemPtr &item)
             _activityModel->addIgnoredFileToList(activity);
         } else {
             // add 'protocol error' to activity list
+            if (item->_status == SyncFileItem::Status::FileNameInvalid) {
+                displayNotification(item->_file, activity._subject);
+            }
             _activityModel->addErrorToActivityList(activity);
```
Comment (allexzander) [inline src/gui/tray/UserModel.cpp:505]: @FlexW Do you still need to call **addErrorToActivityList** if the previous condition is **true**?

Comment (allexzander) [reply src/gui/tray/UserModel.cpp:505]: @FlexW Or, is **displayNotification** only displays a notification to the OS native popup? Like, near tray on Windows? In that case, I'd rename this function to something more meaningful. Like, **showOSNotificationPopup**. Or, whatever, as long as it is clear from the function's name that it's triggering the OS to display a notification via native UI.

Comment (FlexW) [reply src/gui/tray/UserModel.cpp:505]: @allexzander It's still necessary. displayNotification just displays os notifications. I would then go for `showOsNotification()`. Is that ok for you?

File: src/gui/tray/UserModel.cpp [hunk 4]
```diff
@@ -58,52 +59,54 @@ User::User(AccountStatePtr &account, const bool &isCurrent, QObject *parent)
     connect(_activityModel, &ActivityListModel::sendNotificationRequest, this, &User::slotSendNotificationRequest);
 }
 
-void User::slotBuildNotificationDisplay(const ActivityList &list)
+void User::showOsNotification(const QString &title, const QString &message)
```
Comment (er-vin) [inline src/gui/tray/UserModel.cpp:62]: s/Os/Desktop/ ?

These two letter acronyms stuck between other words are always harder to spot. :-)

File: src/gui/tray/UserModel.cpp [hunk 5]
```diff
@@ -58,52 +59,54 @@ User::User(AccountStatePtr &account, const bool &isCurrent, QObject *parent)
...
+{
+    showOsNotification(notificationId(title, message), title, message);
+}
+
+void User::showOsNotification(int id, const QString &title, const QString &message)
```
Comment (er-vin) [inline src/gui/tray/UserModel.cpp:67]: ditto

File: src/gui/tray/UserModel.cpp [hunk 6]
```diff
@@ -496,11 +499,26 @@ void User::slotItemCompleted(const QString &folder, const SyncFileItemPtr &item)
...
 
+int User::notificationId(const QString &subject, const QString &message)
+{
+    const QString idKey = subject + message;
+    const auto idsIter = _notificationIds.find(idKey);
+    if (idsIter != _notificationIds.end()) {
+        return *idsIter;
+    }
+    const auto newId = _notificationIds.size();
+    _notificationIds[idKey] = newId;
```
Comment (er-vin) [inline src/gui/tray/UserModel.cpp:518]: Hm... this will grow "infinitely" over time. At least as long as you get variations on subject and message. Since it contains full concatenation of both... Also said id is kind of fragile and will break as long as you put in place strategies to remove items from that QHash (since you'll start recycling ids).

It needs to be rethought I think. For a start, why did you drop the id member of the activity object? I'd presume it was a better id to start with? Or is the issue that it changes all the time?

If really you need a way to compute more robust numeric ids out of subject and message, I'd advise using the qHash function on this idKey, then you probably don't even need to store it. That raises the question on how often that'd collide though since it's a qint16 if my memory serves me well. I'd really keep that as last resort option though if the id field in the activity objects somehow don't cut it.

Comment (FlexW) [reply src/gui/tray/UserModel.cpp:518]: The reason why I don't use the activity id is that this id comes from the json response of the activities that the server sent to the client. The file name warnings have obviously no such id. So I need to generate a artificial id for them. Also I need to make sure that the ids of the activities and the ids of the file name warnings do not collide. Original this whole notification cache mechanism was only used for the activities that the server sent so it was fine to use the activity id for notification caching.

I'm not that satisfied with that solution too.

**Comment by er-vin** [comment]
> The reason why I don't use the activity id is that this id comes from the
> json response of the activities that the server sent to the client. The
> file name warnings have obviously no such id. So I need to generate a
> artificial id for them. Also I need to make sure that the ids of the
> activities and the ids of the file name warnings do not collide. Original
> this whole notification cache mechanism was only used for the activities
> that the server sent so it was fine to use the activity id for notification
> caching.
> 
> I'm not that satisfied with that solution too.

Ah fair enough. But somehow that's good news. I'd use the id from the activity 
if it is present and derive one from the content if not. Also I'd really try 
hard to avoid storing them. That will require changing the type stored in 
_guiLoggedNotifications since in a way you'd have "two type of ids". 
Interestingly it needs changing already anyway (int vs qlonglong...).

**Comment by FlexW** [comment]
>  I'd use the id from the activity if it is present and derive one from the content if not. 

But if we use the id from the activities AND a self generated, then they might collide, or do you think the risk is small enough?
How would you then generate the artificial id? Would you use the hash value of the content as id?

Another thing: You said that it will grow really big over time but I somehow assumed `QHash` does just store the hash of the string and not the whole string. If `QHash` really stores the string then we could also calculate the hash of the string on our own and store just the hash. The hash value of the string is just a 32 bit number and the ids of the activities were in the past already stored. So I wouldn't expect that the memory consume grows that much. Of course it will need more memory than before because it needs to store the id the hash and the activity id. Do I have a mistake in my thinking?

**Comment by er-vin** [comment]
> But if we use the id from the activities AND a self generated, then they
> might collide, or do you think the risk is small enough?

Oh they *will* collide. That's why I mentioned the type might need adjusting, 
you might want two qlonglong or such (one being used for activity id, other 
one for content based id).

> How would you then generate the artificial id? Would you use the hash value
> of the content as id?

I'm concerned qHash might not have enough entropy indeed... Something else 
might be better in that case. Maybe like QCryptographicHash with a cheap 
algorithm.

To do it properly someone will have to run the math on all that anyway. :-)

> Another thing: You said that it will grow really big over time but I somehow
> assumed `QHash` does just store the hash of the string and not the whole
> string. If `QHash` really stores the string then we could also calculate
> the hash of the string on our own and store just the hash. The hash value
> of the string is just a 32 bit number and the ids of the activities were in
> the past already stored. So I wouldn't expect that the memory consume grows
> that much. Of course it will need more memory than before because it needs
> to store the id the hash and the activity id. Do I have a mistake in my
> thinking?

Well there's more bookkeeping due to the buckets and so on but yes it's not 
that huge per entry *if* it doesn't collide (but it's using qHash internally 
so....). The bigger issue is more that people can keep the client running for 
weeks or months uninterrupted (hello suspend to ram). So it's doomed to grow 
if it's unbounded.

File: src/gui/tray/UserModel.cpp [hunk 7]
```diff
@@ -104,7 +93,7 @@ void User::slotBuildNotificationDisplay(const ActivityList &list)
             continue;
         }
         const auto message = AccountManager::instance()->accounts().count() == 1 ? "" : activity._accName;
-        showOsNotification(activity._subject, message);
+        showDesktopNotification(activity._subject, message);
```
Comment (allexzander) [inline src/gui/tray/UserModel.cpp:96]: @FlexW To mitigate risks of having a collision, I would also think of adding **_accName** into the hash calculation. For those cases when a user has multiple accounts added.

File: src/gui/tray/NotificationCache.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+namespace OCC {
+
+struct Notification
+{
+    QString title;
+    QString message;
+};
+
+class NotificationCache : public QObject
```
Comment (allexzander) [inline src/gui/tray/NotificationCache.h:14]: @FlexW Are you sure you need that (small, but still some) overhead of deriving from QObject? Unless you are using features like `signals/slots` or others(from QMetaObject), I'd say, this is not necessary.

File: src/gui/tray/NotificationCache.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    explicit NotificationCache(QObject *parent = nullptr);
+
+    bool contains(const Notification &notification) const;
+
+    void insert(const Notification &notification);
+
+    void clear();
+
+private:
+    std::set<uint> _notifications;
```
Comment (allexzander) [inline src/gui/tray/NotificationCache.h:28]: @FlexW Just curious, why not use `QSet<uint>`?

File: src/gui/tray/NotificationCache.h [hunk 10]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <set>
+#include <QObject>
+
+namespace OCC {
+
+struct Notification
```
Comment (allexzander) [inline src/gui/tray/NotificationCache.h:8]: @FlexW I guess this struct should be nested into `NotificationCache` just to avoid possible namespace collisions in the future. Also, it feels like this struct is a part of `NotificationCache`.

File: src/gui/tray/UserModel.cpp [hunk 11]
```diff
@@ -58,52 +59,43 @@ User::User(AccountStatePtr &account, const bool &isCurrent, QObject *parent)
...
 {
-    // Whether a new notification was added to the list
-    bool newNotificationShown = false;
+    ConfigFile cfg;
+    if (!cfg.optionalServerNotifications() || !isDesktopNotificationsAllowed()) {
+        return;
+    }
 
+    // after one hour, clear the gui log notification store
+    if (_guiLogTimer.elapsed() > 60 * 60 * 1000) {
```
Comment (allexzander) [inline src/gui/tray/UserModel.cpp:70]: @FlexW Just as a nitpick, you can make this `60 * 60 * 1000` a `constexpr` variable in this function, just for it to be more readable.

File: src/gui/tray/UserModel.cpp [hunk 12]
```diff
@@ -58,52 +59,43 @@ User::User(AccountStatePtr &account, const bool &isCurrent, QObject *parent)
...
+        _notificationCache.clear();
+    }
+
+    const Notification notification { title, message };
+    const auto isNotificationCached = _notificationCache.contains(notification);
```
Comment (allexzander) [inline src/gui/tray/UserModel.cpp:75]: @FlexW This variable seem redundant to me :)

File: src/gui/tray/NotificationCache.h (hunk 9 above)
Comment (FlexW) [reply src/gui/tray/NotificationCache.h:28]: Because we had the discussion and I thought that we use STL whenever possible :)

File: src/gui/tray/UserModel.cpp (hunk 11 above)
Comment (FlexW) [reply src/gui/tray/UserModel.cpp:70]: Yes that would be good :) However, it was like that before.

File: src/gui/tray/UserModel.cpp (hunk 12 above)
Comment (FlexW) [reply src/gui/tray/UserModel.cpp:75]: Yes true.

**Comment by FlexW** [comment]
@er-vin I implemented a simple notification cache that gets cleared after one hour. That should prevent the memory from growing if the app runs for a long time. The behaviour should be consistent with what was there before.

File: src/gui/tray/NotificationCache.cpp [hunk 13]
```diff
@@ -0,0 +1,28 @@
+#include "NotificationCache.h"
+#include <qhash.h>
+
+namespace OCC {
+
+NotificationCache::NotificationCache(QObject *parent)
+    : QObject(parent) {};
```
Comment (er-vin) [inline src/gui/tray/NotificationCache.cpp:7]: This semicolon shouldn't be here. Also each curly brace should be on its own line.

File: src/gui/tray/NotificationCache.h (hunk 8 above)
Comment (er-vin) [reply src/gui/tray/NotificationCache.h:14]: Clearly overkill to inherit from QObject indeed.

File: src/gui/tray/NotificationCache.h (hunk 9 above)
Comment (er-vin) [reply src/gui/tray/NotificationCache.h:28]: By moving away from QObject you might want to move to QSet though. Because right now it doesn't make much difference to use one or the other (QObject not being copyable), but without QObject then it's nice to have the implicit sharing and not have to enter into dealing with efficient copies.

I don't think "STL whenever possible" is a good guideline to be honest. Same thing for "Qt containers whenever possible" actually. As can be seen here they are different enough to both have their own pros and cons.

File: src/gui/tray/NotificationCache.h [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    bool contains(const Notification &notification) const;
+
+    void insert(const Notification &notification);
+
+    void clear();
+
+private:
+    std::set<uint> _notifications;
+
+    uint calculateKey(const Notification &notification) const;
```
Comment (er-vin) [inline src/gui/tray/NotificationCache.h:30]: nitpick: generally better to have the methods come first and then the variables in the private section

File: src/gui/tray/UserModel.cpp (hunk 7 above)
Comment (er-vin) [reply src/gui/tray/UserModel.cpp:96]: This is a good point, actually it occurred to me that the whole approach based on formerly QSet<int> and now set<uint> is in fact weak behavior wise. It's doomed to generate collisions, if you look at qHash implementation it's a hash function to try to balance the buckets in a hash map, it's not something stronger about "do I already have seen this exact string before". With a QSet you're oblivious to such collisions and so the chances are high that some messages are being eaten up even today.

Now it's probably not the role of that PR to fix this, It's clearly too big already this was supposedly about adding another warning type and it's now going to refactor the whole notification handling in here. Would have benefited from being two PRs (one for said warning, and another to rework the desktop notification caching). There's plenty of time in the 3.3 cycle to do this and backport one or both of them on stable-3.2.

But clearly with the discussions around that topic that we had on this PR, the whole use of QSet/set is something to keep an eye on and *maybe* fix. I mean in a way people hate desktop notifications it's not like they're going to shout if some are not happening... OK... apart the ones from Talk those they should always fire or complaints will happen.

File: src/gui/tray/UserModel.cpp [hunk 15]
```diff
@@ -436,33 +425,17 @@ bool User::isUnsolvableConflict(const SyncFileItemPtr &item) const
     // We just care about conflict issues that we are able to resolve
     return item->_status == SyncFileItem::Conflict && !Utility::isConflictFile(item->_file);
 }
-
-void User::slotItemCompleted(const QString &folder, const SyncFileItemPtr &item)
+void User::processCompletedSyncItem(const Folder *folder, const SyncFileItemPtr &item)
```
Comment (er-vin) [inline src/gui/tray/UserModel.cpp:428]: Missing empty line before this one.

**Comment by FlexW** [comment]
@allexzander @er-vin After all the discussions I'm now unsure how to move on. I think at this point the notification mechanism is not weaker than it was before (leaving away theoretical aspects and storing the account as @allexzander mentioned). What should I now do with that pr? Go back to how it was, just generate the id for activities and warnings in a different way and leave the rest the same? Or even leave the activities completely untouched and just store warning ids separate?

I noticed (while working in a different area) that the notification api that we use on MacOS is deprecated and will be removed in the next versions. I think the whole notification system needs a overhaul and maybe we even want to do that notification caching in a more central place so that other parts of the application can use it too and make the code easier to understand. Also we should discuss then upfront on how we want to generate the ids etc so that I not implement something and then need to redo that again because it goes in the wrong direction.

I also wanted to mention that this new `processCompletedSyncItem()` method is a left over from my experiment to introduce unit tests for that part of the application. It turned out that this is at the moment very difficult to achieve because of the monolithic `FolderMan` and `Folder`. Changing that will involve some big refactoring I guess. But maybe we want to invest that time to make our life easier in the long term (testing by hand every time one makes a change costs a lot of time).

**Comment by er-vin** [comment]
> @allexzander @er-vin After all the discussions I'm now unsure how to move
> on.

I see two ways forward. See below.

> I think at this point the notification mechanism is not weaker than it
> was before (leaving away theoretical aspects and storing the account as
> @allexzander mentioned).

Totally agree. My comments in that department were more for "hey, there's 
likely another issue which was there before, file it somewhere and make sure 
it's revisited at some point". It's by no mean a deal breaker for that PR.

> What should I now do with that pr? Go back to how
> it was, just generate the id for activities and warnings in a different way
> and leave the rest the same? Or even leave the activities completely
> untouched and just store warning ids separate?

So, two ways forward I'd say:
 * either complete it almost as is (I guess there's still a couple of smaller 
unaddressed comments), it's went in more directions than I'd like (obviously, 
already voiced this kind of concerns);
 * refocus it by moving back on the "easy version" where it's really about 
introducing the new warning.

First way forward allows to salvage all your work for sure, avoids the 
spamming on the new warning as soon as its merged (I think that was the main 
concern for going down this particular rabbit hole). But... messier history 
(one PR and all in one commit) and adding some unfinished facilities to what's 
already there.

Second way forward means cleaner history and avoids piling half finished 
facilities. But... doesn't prevent the spamming on that new warning type 
(IIRC) and means some rework to move the non cache related changes to another 
PR (or they'll get lost).

Either way... Some extra work will be needed. So I guess it depends which one 
you feel most confident with.

> I noticed (while working in a different area) that the notification api that
> we use on MacOS is deprecated and will be removed in the next versions.

That'd be a good reason to switch to KNotifications by the way. Which goes 
back to: build system fixes.

> I think the whole notification system needs a overhaul and maybe we even
> want to do that notification caching in a more central place so that other
> parts of the application can use it too and make the code easier to
> understand.
> Also we should discuss then upfront on how we want to generate the ids etc
> so that I not implement something and then need to redo that again because
> it goes in the wrong direction.

Definitely, it's just odd currently the code path it has... I mean at some 
point it goes through Logger, gets reemitted, then goes in the systray which 
then suddenly contains mac native code.

> I also wanted to mention that this new `processCompletedSyncItem()` method
> is a left over from my experiment to introduce unit tests for that part of
> the application. It turned out that this is at the moment very difficult to
> achieve because of the monolithic `FolderMan` and `Folder`. Changing that
> will involve some big refactoring I guess. But maybe we want to invest that
> time to make our life easier in the long term (testing by hand every time
> one makes a change costs a lot of time).

And moving to libsync, and the awful AccountManager / AccountState which 
suffer from similar design issue while adding the feat of being very much tied 
to the Application object. Should I also throw owncloudGui in the hat? :-)

Not to diminish your stance in any way. What you mentioned needs to happen at 
some point and what I mentioned as well.

The point I'm getting at here is: as software developers we tend to jump on 
this kind of "hey look at that design bit, it sucks, let's make a clean 
cathedral out of it". It's a bias to keep in check because (especially in such 
a code base) it tends to link to a large amount of other topics (just that 
notification thing will lead you to touch several layers of GUI, FolderMan/
Folder well... pretty much everywhere in the GUI).

You can quickly end up rocking the boat so much that releases get delayed a 
lot. All the while there's a much more boring job of the build system and 
dependency management... but that'd unlock quite a few possibilities to make 
some things easier (that's why I mentioned KNotifications above, I'm not 
married with that one, there are others... point is for this kind of 
commodities you likely want to avoid writing native code yourself to focus on 
your actual business).

In other words, in term of priorities it's about finding the main bottleneck 
and tackling that first. To me currently it's mostly the "build tools" which 
are holding things back (old and somewhat messy cmake code, no build 
dependency management a la conan or KDE Craft, lack of Windows CI, no "per PR 
Windows build", etc.).

Alright, that ended up longer than anticipated, I'll stop here. ;-)

**Comment by FlexW** [comment]
> So, two ways forward I'd say:

I will think about them until tomorrow and decide then. Thanks for taking the time :)

> That'd be a good reason to switch to KNotifications by the way. Which goes
back to: build system fixes.

+1

> And moving to libsync, and the awful AccountManager / AccountState which
suffer from similar design issue while adding the feat of being very much tied
to the Application object. Should I also throw owncloudGui in the hat? :-)

That's true. As we're at it: I don't get the line between libsync and gui. There are a few things in libsync that should be in gui in my opinion. Things like `theme.cpp`.

> To me currently it's mostly the "build tools" which
are holding things back (old and somewhat messy cmake code, no build
dependency management a la conan or KDE Craft, lack of Windows CI, no "per PR
Windows build", etc.).

I fully agree with that.

File: src/gui/tray/NotificationCache.h (hunk 14 above)
Comment (FlexW) [reply src/gui/tray/NotificationCache.h:30]: I change it as we do it always that way, but in general I think it has benefits to have the data before the method. If I want to understand a piece of code, I always look at first at the data because the data somehow defines the class, and the goal of every operation that we do with the computer is to change some data.

**Comment by FlexW** [comment]
@er-vin @allexzander I decided to let the new notification cache in. It is not perfect but it makes it easier to add other kind of warnings in the near future. I'm not sure how long it will take until we clean that code up. I also think it's not necessary to save any kind of account information in the notification cache as the notification cache is already unique per account. Other than that, I think I addressed all other comments.

File: src/gui/tray/UserModel.h [hunk 16]
```diff
@@ -91,17 +94,21 @@ public slots:
...
     bool isUnsolvableConflict(const SyncFileItemPtr &item) const;
 
+    void showDesktopNotification(const QString &title, const QString &message);
+
 private:
     AccountStatePtr _account;
     bool _isCurrentUser;
     ActivityListModel *_activityModel;
     ActivityList _blacklistedNotifications;
+    // QHash<QString, int> _notificationIds;
```
Comment (er-vin) [inline src/gui/tray/UserModel.h:104]: Probably want to remove the commented out code now

**Review by er-vin — COMMENTED** [review]
Just some commented out code left behind, this needs clean up. History as well needs a bit of a clean up, bonus point for keeping the NotificationCache introduction in its own commit. :-)

**Comment by FlexW** [comment]
@er-vin Should be ready now. Thanks for your help :)

**Review by er-vin — COMMENTED** [review]
LGTM, thanks

**Review by allexzander — APPROVED** [review]
Ok. @FlexW We just need to improve the uniqueness of hashed values by adding **_accName** as mentioned in one of my comments, and, proposed by @er-vin to do it in a separate PR. Agree with that.

**Comment by FlexW** [comment]
> Ok. @FlexW We just need to improve the uniqueness of hashed values by adding **_accName** as mentioned in one of my comments, and, proposed by @er-vin to do it in a separate PR. Agree with that.

Are you sure this would bring any value? Because the notification cache is already tied to the User object, and as far as I understand there is one User object per account. So it is no problem at the moment if the same message arrives on different accounts.

**Comment by allexzander** [comment]
> > Ok. @FlexW We just need to improve the uniqueness of hashed values by adding **_accName** as mentioned in one of my comments, and, proposed by @er-vin to do it in a separate PR. Agree with that.
> 
> Are you sure this would bring any value? Because the notification cache is already tied to the User object, and as far as I understand there is one User object per account. So it is no problem at the moment if the same message arrives on different accounts.

@FlexW Ah yes, ok then, never mind.

**Comment by FlexW** [comment]
> > > Ok. @FlexW We just need to improve the uniqueness of hashed values by adding **_accName** as mentioned in one of my comments, and, proposed by @er-vin to do it in a separate PR. Agree with that.
> > 
> > 
> > Are you sure this would bring any value? Because the notification cache is already tied to the User object, and as far as I understand there is one User object per account. So it is no problem at the moment if the same message arrives on different accounts.
> 
> @FlexW Ah yes, ok then, never mind.

NP :)
